### DISCUSSION TITLE
using model matrix or view matrix in the shader

### DISCUSSION THREAD

**Discussion by MasonAlice** [vsg-dev/VulkanSceneGraph#693] [General]
I figured out that vsg use push constant to update projection matrix and modelview matrix. how can I just use model or view matrix in the shader.

**Comment by robertosfield** [comment]
If you can avoid using view and model matrices in shaders you can avoid precision problems that occur when the view matrix values become large.  The VulkanSceneGraph is designed to handle really large to tiny worlds where ability handle precision issues are critical so use modelview matrix are the default.  

Sadly lots of example shaders out there use separate model and view matrices are introduce this precision problem, then when developers trying to scale there worlds from toy examples into large real world scales they are hit by an intractable precision problems.

While it's technically possible to pass separate model and view matrices directly yourself, I would strongly recommend thinking about how to refactor shaders so they can work with modelview matrices that the VSG provides out of the box.  We can help make suggestions, but wihtout knowing what might be in your shaders and what type of scene you are trying to render it's too open ended attempt.

**Comment by robertosfield** [comment]
I'm moving this "Issue" to the Discussion pages where it belongs as it's not a bug or build problem with the VSG itself.

**Comment by timoore** [comment]
I was going to reply with essentially the same thing that @robertosfield said. If you really need these matrices, then the most productive way to proceed would be to modify  vsg::ViewDependentState to (optionally) pass the view matrix and view matrix inverse in its uniform buffer. Once you have the those, you can extract the model matrix from the ModelView matrix. Robert can comment on whether or not he would accept such a patch.

**Reply by robertosfield** [reply]
If one just passes the view matrix then the vertex shader would have to compute the model matrix by inverting the view matrix and then multplying against the modelview matrix to get the model matrix.

However, all this extra work to give a use a system that causes precision problems.  It's shooting yourself in the foot because you don't know any better.

The better in this instance is to not use shaders that use separate view and model matrices.  OpenGL back the eighties knew about how handle precision problems properly and just had a modelview matrix stack, and put lights and texgen in eye coordinates in order to lighting and texgen calculations on the GPU.  This system wasn't broke, it was the right way to get graphics systems scale to large coordinate models.

So I would prefer the path of education rather than adding a kludge into the core VSG so that users can shoot themselves in the foot and then later waste lots of time on precision issues because the foot was so numb they didn't even realize they shot themselves in the foot.

**Reply by robertosfield** [reply]
W.r.t ViewDependentState - it exists to place data like postions used in the shaders into eye coordinates, just like OpenGL fixed function pipeline does.  The PBR/Phong shaders that the VSG provides all do their light calcs in eye space rather than world coords specifically to enable us to seamlessly scale to whole earth rendering.

**Reply by MasonAlice** [reply]
good explanation!

**Reply by robertosfield** [reply]
I think it be would having  an example/set of examples that illustrate how best to do different features light lighting, texgen, handling large worlds and how to re-interpret 3rd party shaders that use separate view and model matrices in the way that works well with the approach the VSG provides.  While I know the implementation details conveying this all is the missing bit of jigsaw.
